lib/fog/aws/models/storage/file.rb

```diff
@@ -460,6 +476,49 @@ def upload_in_threads(target_directory_key, target_file_key, upload_id, pending,
 
           threads.map(&:value).compact
         end
+
+        def upload_parts_in_threads(directory_key, target_file_key, upload_id, pending, completed, thread_count)
+          mutex = Mutex.new
+          threads = []
+
+          thread_count.times do
+            thread = Thread.new do
+              begin
+                while part = pending.shift
+                  # Determine byte range for this part
+                  start_pos = (part.part_number - 1) * multipart_chunk_size
+
+                  # Safely read the chunk from body
+                  chunk = nil
+                  mutex.synchronize do
+                    if body.respond_to?(:seek)
+                      body.seek(start_pos) rescue nil
```

Comment:
hello, it is me, a little bottleneck right here!

Reply:
Yeah. This should prevent multiple threads from trying to read the file at the same time I guess. They would take turns leading it into memory, but then actually do the upload in parallel (which is the slower part). I think mutexing the file read is probably necessary if I recall properly. Is that what you were thinking? Happy to discuss further.

Reply:
Yes, this will definitely be a bottleneck on a block storage side. Multiple threads trying to seek through blocks and run IO operations from one file, that is a reason why mutex is there.

Reply:
Any ideas for alternatives or workarounds?

Reply:
in this implementation it's implied that we could fine-tune this bottleneck using two existing properties, `concurrency` and `multipart_chunk_size`

Reply:
Using some sort of buffering in addition to threads wouldn't make it any better.